packages/ui/src/Tooltip.tsx

```diff
@@ -1,37 +1,122 @@
+  children: (ref: React.Dispatch<React.SetStateAction<HTMLElement | null>>) => ReactNode
 }
 
-export const Tooltip: FC<TooltipProps & { children: ReactElement }> = ({
-  overlay,
-  wrapMaxWidth = true,
-  wrapElement = false,
+/**
+ * ### Purpose
+ * Useful when you need to display additional information/actionable content.
+ * A tooltip with this content is displayed when user hovers over a target element.
+ *
+ * ### General info
+ * - Tooltip automatically detects viewport overflow and changes placement to prevent it.
+ * - Offset (space between a target and tooltip elements) can be also configured via "offset" property.
+ * - If "content" property is undefined, tooltip element will be removed from DOM entirely.
+ * - It's required to set the "id" property which will be assigned to the tooltip. This id parameter can be then used as a value of "aria-labelledby" attribute.
+ *
+ * ### Usage
+ * Wrap the target element with Tooltip component and use the "content" property to define what should be displayed inside the tooltip.
+ */
+export const Tooltip: FC<TooltipProps> = ({
+  content,
+  hideTimeoutDuration = 100,
+  offset = 10,
+  placement = 'top',
+  visible: visibilityOverride,
   children,
   ...props
 }) => {
-  return overlay ? (
-    <RCTooltip
-      overlay={wrapMaxWidth ? <div className={styles.wrapper}>{overlay}</div> : overlay}
-      overlayClassName={styles.overlay}
-      {...props}
-    >
-      {wrapElement ? (
-        // TODO: remove once https://github.com/react-component/tooltip/issues/18#issuecomment-411476678 has been resolved
-        <div style={{ display: 'inline-block' }}>{children}</div>
-      ) : (
-        children
+  const [isShown, setShown] = useState<boolean>(false)
+  const [hideTimeout, setHideTimeout] = useState<NodeJS.Timeout | null>(null)
+
+  /*
+   * We're using useState instead of useRef because react-popper package has some issues with useRef:
+   * https://github.com/popperjs/react-popper/issues/241#issuecomment-591411605
+   */
+  const [referenceElement, setReferenceElement] = useState<HTMLElement | null>(null)
+  const [popperElement, setPopperElement] = useState<HTMLSpanElement | null>(null)
+  const [arrowElement, setArrowElement] = useState<HTMLSpanElement | null>(null)
+
+  const { styles: popperStyles, attributes: popperAttributes } = usePopper(referenceElement, popperElement, {
+    placement,
+    modifiers: [
+      {
+        name: 'arrow',
+        options: {
+          element: arrowElement,
+        },
+      },
+      {
+        name: 'offset',
+        options: {
+          offset: [0, offset],
+        },
+      },
+    ],
+  })
+
+  const onMouseEnter = useCallback(() => {
+    if (hideTimeout !== null) {
+      clearTimeout(hideTimeout)
+    }
+
+    setShown(true)
+  }, [hideTimeout])
+
+  const onMouseLeave = useCallback(() => {
+    setHideTimeout(setTimeout(() => setShown(false), hideTimeoutDuration))
+  }, [hideTimeoutDuration])
+
+  useEffect(() => {
```

Comment:
Should we remove the event listeners when they are no longer needed?

Reply:
Updated 👍